crates/kubelet/src/volume/mod.rs

```diff
@@ -243,3 +158,78 @@ impl From<Option<String>> for ItemMount {
         }
     }
 }
+
+/// This is a gnarly function to check all of the supported data members of the
+/// Volume struct. Because it isn't a HashMap, we need to check all fields
+/// individually
+async fn configure(
+    vol: &KubeVolume,
+    namespace: &str,
+    client: &kube::Client,
+    plugin_registry: Arc<PluginRegistry>,
+    path: &PathBuf,
+) -> anyhow::Result<Type> {
+    if let Some(cm) = &vol.config_map {
+        let name = &cm
+            .name
+            .as_ref()
+            .ok_or_else(|| anyhow::anyhow!("no configmap name was given"))?;
+        let cm_client: Api<ConfigMap> = Api::namespaced(client.clone(), namespace);
+        let config_map = cm_client.get(name).await?;
+        configmap::populate(config_map, path, &cm.items).await
+    } else if let Some(s) = &vol.secret {
+        let name = &s
+            .secret_name
+            .as_ref()
+            .ok_or_else(|| anyhow::anyhow!("no secret name was given"))?;
+        let secret_client: Api<Secret> = Api::namespaced(client.clone(), namespace);
+        let secret = secret_client.get(name).await?;
+        secret::populate(secret, path, &s.items).await
+    } else if let Some(pvc_source) = &vol.persistent_volume_claim {
+        debug!("found a PVC source");
+        let pvc_client: Api<PersistentVolumeClaim> = Api::namespaced(client.clone(), namespace);
+        let pv_client: Api<PersistentVolume> = Api::all(client.clone());
+        let storage_class_client: Api<StorageClass> = Api::all(client.clone());
+
+        let pvc = pvc_client.get(&pvc_source.claim_name).await?;
+        debug!("found a PVC");
+        let spec = match &pvc.spec {
+            Some(s) => s,
+            None => {
+                return Err(anyhow::anyhow!("PersistentVolumeClaim must specify a spec"));
+            }
+        };
+        persistentvolumeclaim::validate(&spec)?;
+
+        let storage_class = storage_class_client
+            // NOTE(bacongobbler): should be safe to unwrap() here after calling
+            // validate(). Storage class names are required as we do not define
+            // a default storage class.
+            //
+            // If we do implement a default storage class, then we'll have to
+            // revisit this assumption (omission of this field implies the
+            // default).
+            //
+            // https://kubernetes.io/docs/concepts/storage/persistent-volumes/#class-1
+            .get(spec.storage_class_name.as_ref().unwrap())
+            .await?;
+        debug!("found a storage class");
+        // TODO(bacongobbler): use something a bit more gentle than `.expect`
+        let volume_name = spec.volume_name.as_ref().expect("volume name must exist");
+        // TODO(bacongobbler): When a PVC specifies a selector in addition to
```

Comment:
Isnt this handled by k8s when the PVC is bound?

Reply:
If I can recall from memory... The volume manager code selects and determines the correct storage class. This is performed at runtime. Kubernetes does not handle this for the Kubelet.

I'll see if I can find the relevant code bits again and drop a doc comment to give some more context around this.